server/src/main/java/org/opensearch/index/ShardIndexingPressureTracker.java

```diff
@@ -0,0 +1,211 @@
+ * b. totalBytes - Total bytes that are processed/completed successfully for a shard.
+ * c. requestCount - Total number of requests that are processed/completed successfully for a shard.
+ *
+ * 2. RejectionTracker : To track the rejection statistics for a shard
+ * a. totalRejections - Total number of requests that were rejected for a shard.
+ * b. nodeLimitsBreachedRejections - Total number of requests that were rejected due to the node level limits breached
+ *          i.e. when a request for a shard could not be served due to node level limit was already reached.
+ * c. lastSuccessfulRequestLimitsBreachedRejections - Total number of requests that were rejected due to the
+ *          last successful request limits breached for a shard i.e. complete path failure (black-hole).
+ * d. throughputDegradationLimitsBreachedRejections - Total number of requests that were rejected due to the
+ *          throughput degradation in the request path for a shard i.e. partial failure.
+ *
+ * 3. Performance Tracker : To track the request performance statistics for a shard
+ * a. latencyInMillis - Total indexing time take by requests that were processed successfully for a shard.
+ * b. lastSuccessfulRequestTimestamp - Timestamp of last successful request for a shard.
+ * c. TotalOutstandingRequests - Total requests outstanding for a shard at any given point.
+ * d. ThroughputMovingAverage - Total moving average throughput value for last N requests.
+ * e. ThroughputMovingQueue - Queue that holds the last N requests throughput such that there exists a sliding window
+ *          which keeps moving everytime a new request comes. At any given point it tracks last N requests only.
+ *          EWMA cannot be used here as it evaluate the historical average
+ *          while here it just needs the average of just last N requests.
+ *
+ * ShardIndexingPressureTracker therefore provides the construct to track all the write requests targeted for a ShardId on the node,
+ * across all possible transport-actions i.e. Coordinator, Primary and Replica.
+ * Tracker is uniquely identified against a Shard-Id on the node. Currently the knowledge of shard roles (such as primary vs replica)
+ * is not explicit to the tracker, and it is able to track different values simultaneously based on the interaction hooks of the
+ * operation (action) layers.
+ *
+ * There is room for introducing more unique identity to the trackers based on Shard-Role or Shard-Allocation-Id, but that will also
+ * increase the complexity of handling shard-lister events and handling other scenarios such as request-draining etc.
+ *
+ * To prefer simplicity we have modelled by keeping explicit fields for different operation tracking, while tracker by itself is
+ * agnostic of the role.
+ */
+public class ShardIndexingPressureTracker {
+
+    private final ShardId shardId;
+    private final AtomicLong primaryAndCoordinatingLimits;
+    private final AtomicLong replicaLimits;
```

Comment:
Are these limits going to change if yes you'd want to label them as `initial*Limits`

Reply:
These reflect the shard upper limits (memory) for primaryAndCoordinating and replica roles respectively. They are expected to be updated throughout the course of lifecycle of tracker as when more quota is granted or released.

Reply:
I meant in the constructor.

Reply:
ohh okay. Makes sense. Renamed.